internal/contentmanager/contentmanager.go

```diff
@@ -5,219 +5,331 @@ import (
-// It utilizes the provided ClusterExtension to set a DefaultLabelSelector on the cache.Cache
-// to ensure it is only caching and reacting to content that belongs to the ClusterExtension.
-// For each client.Object provided, a new source.Kind is created and used in a call to the Watch() method
-// of the provided controller.Controller to establish new watches for the managed resources.
-func (i *instance) Watch(ctx context.Context, ctrl controller.Controller, ce *v1alpha1.ClusterExtension, objs []client.Object) error {
+// CreateOrUpdate creates or updates a managed content cache for the provided ClusterExtension and managed objects.
+// At a very high-level, the logical flow of this function is:
+//   - Calculate the set of GVKs to create sources for
+//   - IF a cache does not already exist, create one
+//   - IF a cache already exists, calculate the difference between sources that have already been created. Identify which new sources need to be created and which need to be stopped and removed.
+//   - Stop and remove any sources that need to be stopped and removed
+//   - Create new sources, filtering the content they receive from the Kubernetes API server with the labels that should be set on content managed by the provided ClusterExtension
+//   - Update and set the cache with the newly created sources
+//
+// An error is returned in the following cases:
+//   - An error occurred while preparing to create the sources. In this case, no sources will have been started and no events on any of the managed content will be reacted to
+//   - An error occurred while syncing a source, or syncing timed out. In this case, any sources that were able to successfully sync will be added to the cache and react to events. All sources that failed to sync will be represented in the returned error
+//
+// In the event a source encounters an error after having previously synced successfully,
+// the source will be stopped and removed. It will no longer react to events.
+func (i *managerImpl) CreateOrUpdate(ctx context.Context, ctrl controller.Controller, ce *v1alpha1.ClusterExtension, objs []client.Object) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
 	if len(objs) == 0 || ce == nil || ctrl == nil {
 		return nil
 	}
 
 	cfg, err := i.rcm(ctx, ce, i.baseCfg)
 	if err != nil {
-		return fmt.Errorf("getting rest.Config for ClusterExtension %q: %w", ce.Name, err)
+		return fmt.Errorf("getting rest.Config: %w", err)
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("getting dynamic client: %w", err)
 	}
 
-	scheme, err := buildScheme(objs)
+	gvkSet, err := gvksForObjects(objs...)
 	if err != nil {
-		return fmt.Errorf("building scheme for ClusterExtension %q: %w", ce.GetName(), err)
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	cache, ok := i.caches[ce.Name]
+	if !ok {
+		cache = NewCache()
+	}
+
+	cacheGvks := sets.New[schema.GroupVersionKind]()
```

Comment:
Verify that these GVKs won't be stale due to removal of GVKs due to watch errors after fetching the set of existing sources.